Handle rule descriptions according to BIG-IP version

edarzins opened the pull request:
Issues:
Fixes #1307

Problem:
The 'description' parameter does not exist in policy rules prior
to BIG-IP v12.1.0. Attempting to set the rule description for these
earlier BIG-IP versions results in an error response.

Analysis:
If the BIG-IP version is earlier than v12.1.0, strip the description from
policy rule.

Tests:
test_rules_description (f5/bigip/tm/ltm/test/functional/test_policy.py)

Commit:
Handle rule descriptions according to BIG-IP version

Issues:
Fixes #1307

Problem:
The 'description' parameter does not exist in policy rules prior
to BIG-IP v12.1.0. Attempting to set the rule description for these
earlier BIG-IP versions results in an error response.

Analysis:
If the BIG-IP version is earlier than v12.1.0, strip the description from
policy rule.

Tests:
test_rules_description (f5/bigip/tm/ltm/test/functional/test_policy.py)

wojtek0806 commented on f5/bigip/tm/ltm/policy.py:
```diff
@@ -94,6 +96,9 @@ def _modify(self, **patch):
         tmos_ver = self._meta_data['bigip']._meta_data['tmos_version']
+        if LooseVersion(tmos_ver) < LooseVersion('12.1.0'):
+            for r in patch.get('rules', []):
+                r.pop('description', None)
```
If we are going to make this change, and what it seems this might be one of many such "exceptions", it would be prudent to have this folded into _meta_data, for example _meta_data['optional_parameters'], similarly to:

https://github.com/F5Networks/f5-common-python/blob/development/f5/bigip/resource.py#L839

Also if we silently remove the description from **kwargs how would the caller know something went wrong? If the caller does not know that a certain attribute does not exist in the specific version, we should inform him by raising exception with appropriate method, rather than just removing.

The only exception to such rule is here, and this is due to the fact that we send the entire __dict__ (minus __meta_data) to the device when we do update(), and state/session parameters are read only in specific situations, so the caller need not to be notified, unless the caller attempt to set them to an illegal value:

https://github.com/F5Networks/f5-common-python/blob/development/f5/bigip/tm/ltm/node.py#L95
https://github.com/F5Networks/f5-common-python/blob/development/f5/bigip/tm/ltm/pool.py#L131

We raise if the parameter is set to illegal value by the caller:
https://github.com/F5Networks/f5-common-python/blob/development/f5/bigip/tm/ltm/pool.py#L209

wojtek0806 replied:
The above behavior of update() will be modified once i get a chance, the issue is already opened up for it:
https://github.com/F5Networks/f5-common-python/issues/1063

wojtek0806 reviewed (changes requested):
Please rework this as per my comments

Commit:
Resolve code-review issues

Created a metadata entry to denote the optional policy parameters
and created a helper function to filter out these version-specific
optionals.

wojtek0806 commented on f5/bigip/tm/ltm/policy.py:
```diff
@@ -56,6 +59,24 @@ def __init__(self, policy_s):
+                    for parm in parms:
+                        value = r.pop(parm, None)
+                        if value is not None:
+                            logger.info(
```
I would suggest raising an error, it is easier to handle by the caller

edarzins replied:
I discussed this with @pjbreaux and @zancas and we agreed that the caller should not receive an exception. SDK should handle the BIG-IP version compatibility issue by stripping the optional parameter and logging the event.

wojtek0806 replied:
This is taken from our dev playbook, which is what we are adhering to, so what has changed? specifying a non existent parameter is a failure

````
Failures generate exceptions
    If the SDK enters a surprising or unexpected state it raises an exception. That's it. It's not generally up to the SDK to implement decision logic that handles edge-cases.. EXCEPT where the SDK is smoothing known issues in the device REST server. (See below.)

````

wojtek0806 commented on f5/bigip/tm/ltm/policy.py:
```diff
@@ -56,6 +59,24 @@ def __init__(self, policy_s):
         self._meta_data['attribute_registry'] = temp
+        self._meta_data['optional_parameters'] = {'rules': ['description']}
+
+    def _filter_version_specific_options(self, tmos_ver, **kwargs):
```
i think the point was missed here, i was talking about streamlining this into resource.py, but fine since this is a bigger job, just open an issue for it and assign it to me, i will do this later with update() rework

wojtek0806 reviewed (changes requested):
i would remove logging and raise exception, after that is done this can be merged

caphrim007 commented:
I look at the description issue as a form of deprecation. Deprecation needs raise no exception. Indeed, the code will continue on, but along the way will need to tweak things because of the deprecated code path

So for now I'm ok with this
